Review sections 3-4. (#113)
mlswg/mls-architecture: draft-ietf-mls-architecture.md

# Authentication Service

The Authentication Service (AS) has to provide three functionalities:

1. Issue credentials to clients that attest to bindings between identities and
   signature key pairs

> 2. Enable a group member to verify that a credential presented by another member

Reviewer: ```suggestion
2. Enable a client to verify that a credential presented by another client
```

   is valid with respect to a reference identifier

3. Enable a group member to verify that a credential represents the same

>    application-level device as another credential

Reviewer: ```suggestion
   application-level device as another credential when possible
```

* A credential from the Authentication Service attesting to the
  binding between the identity and the client's signature key.

* The client's asymmetric encryption public key material.
* The client's asymmetric encryption public key.

All the parameters in the KeyPackage are signed with the signature private key
corresponding to the credential.

As noted above, users may own multiple clients, each with their
own keying material, and thus there may be multiple entries
stored by each user.

The Delivery Service is also responsible for allowing users to add,
remove or update their initial key material, and for ensuring
that the identifier for these keys are unique across all keys

> stored on the Delivery Service.

Author: Removed because this seems to be redundant with paragraph 1 of this section + the last sentence (DS ensures uniqueness) is not true

  which the message has been sent;

* The Epoch number, which represents the number of changes (version) of
  the group associated with a specific group ID, and allows for
  lexicographical ordering of messages from different epochs within the
  same group;

* The Content Type of the message, which allows the Delivery Service to
  determine the ordering requirement on the message, in particular
  distinguishing Commit messages from other messages.

The MLS protocol itself can verify these properties. For instance, if the
Delivery Service reorders messages from a client or provides different clients
with inconsistent orderings, then clients can put messages back in their proper
order.  The asynchronous nature of MLS means that within an epoch, messages are
only ordered per-sender, not globally.

Note that some forms of Delivery Service misbehavior are still possible and
difficult to detect. For instance, a Delivery Service can simply refuse
to relay messages to and from a given client. Without some
sort of side information, other clients cannot generally
distinguish this form of Denial of Service (DoS) attack.

> ## Membership knowledge

Author: Removed this subsection because using MLS in an anonymous way is a very difficult topic and I don't think it's treated fairly here. Using MLS anonymously is more worthy of it's own top-level section with strong guidance.

Group membership is itself sensitive information and MLS is designed
to limit the amount of persistent metadata. However, large
groups often require an infrastructure which provides server fanout.
In the case of client fanout, the destination of a message is known by
all clients, hence the server usually does not need this information.
However, they may learn this information through traffic analysis.
Unfortunately, in a server-side fanout model, the Delivery Service can
learn that a given client is sending the same message to a set of other
clients. In addition, there may be applications of MLS in which the group
membership list is stored on some server associated with the Delivery
Service.

While this knowledge is not a breach of the protocol's authentication or
confidentiality guarantees, it is a serious issue for privacy.  In the case
where metadata has to be persisted for functionality, it SHOULD be stored
encrypted at rest. Applications should also consider anonymous systems for
server fanout such as Loopix {{Loopix}}.

> ## Membership and offline members

Author: Removed this section because it's not related to the DS discussion

For the most part, MLS does not require the Delivery Service to deliver messages
in any particular order. The one exception is that, because an MLS group has a
linear history, the members of the group must agree on the order in which
changes are applied.  Concretely, the group must agree on a single MLS Commit
message that ends each epoch and begins the next one.

In practice, there's a realistic risk of two members generating Commit messages
at the same time, based on the same state, and both attempting to send them to
the group at the same time. The extent to which this is a problem, and the
appropriate solution, depends on the design of the Delivery Service. Per the CAP
theorem, there are two general classes of distributed system:

* Consistent and Partition-tolerant, or CP, systems can provide a globally
  consistent view of data but may stop working if there are network issues.
* Available and Partition-tolerant, or AP, systems continue working despite
  network issues but may return different views of data to different users.

Strategies for sequencing messages in CP and AP systems are described in the
next two subsections.

> However, note that a Delivery Service could also reorder messages or provide an

Reviewer: ```suggestion
However, note that a malicious Delivery Service could also reorder messages or provide an
```
